Merge branch 'main' of https://github.com/svgogin/ServiceRepository into ExceptionHandlers

svgogin opened the pull request:
#20 Add exception handler
1. Exception handler added
2. Server error messages enabled

Commit:
#20 Add exception handler
1. Exception handler added
2. Server error messages enabled

Commit:
Merge branch 'main' of https://github.com/svgogin/ServiceRepository into ExceptionHandlers

 Conflicts:
	src/main/resources/application.properties

alinkov commented on src/main/java/ru/svgogin/service/spark/web/SparkController.java:
```diff
@@ -37,8 +41,13 @@ public SparkController(SparkService sparkService) {
+    if (companyInDb.isEmpty()) {
+      return new ResponseEntity<>(sparkService.save(companyDto), HttpStatus.OK);
+    } else {
+      throw new EntityAlreadyExistsException("Company with inn " + companyDto.getInn() + " already exists");
```
why do you move this logic from service to controller?

svgogin replied:
Fixed

alinkov commented on src/main/java/ru/svgogin/service/spark/exception/EntityAlreadyExistsException.java:
```diff
@@ -0,0 +1,25 @@
+
+  public EntityAlreadyExistsException(String message) {
+    this.message = message;
+    log.error(message);
```
it is not good idea to log here.
It is better if you log it in the `ExceptionHandler`

svgogin replied:
Fixed

alinkov commented on src/main/java/ru/svgogin/service/spark/web/SparkController.java:
```diff
@@ -37,8 +41,13 @@ public SparkController(SparkService sparkService) {
   @PostMapping()
-  public CompanyDto saveCompany(@RequestBody CompanyDto companyDto) {
-    return sparkService.save(companyDto);
+  public ResponseEntity<CompanyDto> saveCompany(@RequestBody CompanyDto companyDto) {
```
It will be good if you add the tests

svgogin replied:
Tests added

alinkov reviewed (commented):
Please fix checkstyle issues

Commit:
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

alinkov commented on src/main/java/ru/svgogin/service/spark/exception/EntityAlreadyExistsException.java:
```diff
@@ -0,0 +1,20 @@
+  private final String message;
+
+  public EntityAlreadyExistsException(String message) {
+    this.message = message;
```
```suggestion
    super(message);
```
And you don't need to override `message` and you can remove field `message`

svgogin replied:
Done

alinkov commented on src/main/java/ru/svgogin/service/spark/service/SparkService.java:
```diff
@@ -43,14 +44,17 @@ public CompanyDto update(CompanyDto companyFromDb, CompanyDto companyDto) {
+    if (companyInDb) {
+      log.info("Company with inn "
+               + companyDto.getInn()
+               + " was not saved. Post method used instead of Put");
```
Please revert to previous message

svgogin replied:
Done

alinkov replied:
Do you push it?

svgogin replied:
There was some misunderstanding. I think it's ok now.

alinkov commented on src/main/java/ru/svgogin/service/spark/service/SparkService.java:
```diff
@@ -43,14 +44,17 @@ public CompanyDto update(CompanyDto companyFromDb, CompanyDto companyDto) {
+                                             + companyDto.getInn()
+                                             + " already exists");
     }
+    log.info("Company with inn " + companyDto.getInn() + " was successfully saved.");
```
You log message before save. It is better to log after real insert

svgogin replied:
I can't swap these lines because 

log.info("Company with inn " + companyDto.getInn() + " was successfully saved.");

becomes an unreachable statement.

So this variant doesn't work.

  public CompanyDto save(CompanyDto companyDto) {
    var companyInDb = sparkRepositoryDb.existsByInn(companyDto.getInn());
    if (companyInDb) {
      log.info("Company with inn "
               + companyDto.getInn()
               + " was not saved. Post method used instead of Put");
      throw new EntityAlreadyExistsException("Company with inn "
                                             + companyDto.getInn()
                                             + " was not saved. Post method used instead of Put");
    }
    return toDto(aggregateTemplate.insert(toEntity(companyDto)));
    log.info("Company with inn " + companyDto.getInn() + " was successfully saved.");
  }

alinkov replied:
```java
var savedCompany = aggregateTemplate.insert(toEntity(companyDto));
log.info("Company with inn " + companyDto.getInn() + " has been saved");
return toDto(savedCompany);
```

svgogin replied:
Done. Thank you. Can you explain why this error occures?

alinkov replied:
which error?

alinkov commented on src/main/resources/migrations/V1__Initial.sql:
```diff
@@ -1,12 +1,12 @@
-    ogrn text,
-    kpp text,
-    full_name_rus text,
+    id             bigint GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
```
nit: you shouldn't update previous migrations

svgogin replied:
This happpened because I used "Reformat code".

svgogin replied:
This happened because I used reformat code for all the packages. I was faced with a hash sum problem. Now I know it, I won't do it again for the whole project.

Commit:
Add exception handler #20
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

Commit:
Add exception handler #20
1. Logic of the Post method was transferred from the controller to the service
2. Tests of the exception handler added.
3. Checkstyle issues fixed.
4. Logging of the logic added to the Post method.

alinkov commented on src/main/java/ru/svgogin/service/spark/service/SparkService.java:
```diff
@@ -43,14 +44,18 @@ public CompanyDto update(CompanyDto companyFromDb, CompanyDto companyDto) {
-      return toDto(aggregateTemplate.insert(company));
+    var companyInDb = sparkRepositoryDb.existsByInn(companyDto.getInn());
+    if (companyInDb) {
+      log.info("Company with inn "
```
you should choose another log lever or it is better if you log it into exception handler
